### PR TITLE
Add StaticInputHandler protocol for zero-allocation input filling

### PR DESCRIPTION
**Pull request by stikves**
Adds `InputBuffers` and `StaticInputHandler`, a protocol for filling pre-allocated input buffers in-place without per-step allocation.

### Design

```swift
public protocol StaticInputHandler: Sendable {
    var inputNames: [String] { get }
    func registerBuffers(into buffers: inout InputBuffers)
    func fill(_ context: InputContext, into buffers: inout InputBuffers) throws
}
```

- Engine owns `InputBuffers`, creates once at init, passes `inout` each step
- `_modify` accessor on `InputBuffers` subscript avoids COW (26× vs get/set)
- Non-mutating `fill` — handler holds no per-step state
- Backward compatible: existing `SyncInputHandler` unchanged

Primary benefit is architectural (no COW risk, compile-time ownership via inout) rather than raw throughput.

### CODE REVIEW

**Review by tjia1818 — APPROVED** [review]

File: swift/Sources/CoreAILanguageModels/Handlers/InputHandler.swift [hunk 1]
```diff
@@ -88,3 +94,88 @@ public enum InputCoverage {
...
+public struct InputBuffers {
+    private var buffers: [String: NDArray] = [:]
+    private var descriptors: [String: NDArrayDescriptor] = [:]
+
+    public init() {}
+
+    /// Register a named input with its descriptor. Allocates initial buffer at batch=1.
+    public mutating func register(name: String, descriptor: NDArrayDescriptor) {
+        descriptors[name] = descriptor
+        let resolved = descriptor.resolvingDynamicDimensions([1, 1])
```
Comment (carinapeng) [inline swift/Sources/CoreAILanguageModels/Handlers/InputHandler.swift]: Hard blocker is here

We get a crash because register is always substituting [1, 1] but static extend inputs have fixed dimensions - position_id is [1, 16] so this would give the error which we saw "Shape at dimension 1 of 1 is not a valid substitution for source shape 16" for Qwen3 0.6b

The fix would be allocating straight from the resolved descriptor, or only substitute dimensions that are actually dynamic (-1) and leave fixed dimensions alone

Also note register hardcodes a 2-D [1, 1] regardless of rank, the causal mask is 4D and step is 1D so it's wrong for those too

File: swift/Sources/CoreAILanguageModels/Handlers/InputHandler.swift [hunk 2]
```diff
@@ -88,3 +94,88 @@ public enum InputCoverage {
...
+        buffers[name] = NDArray(descriptor: resolved)
+    }
+
+    /// Ensure the named buffer has capacity for the given shape.
+    /// Only reallocates if the current shape doesn't match.
+    public mutating func ensureCapacity(name: String, shape: [Int]) {
+        guard let desc = descriptors[name] else { return }
+        let current = buffers[name]
+        let resolved = desc.resolvingDynamicDimensions(shape)
```
Comment (carinapeng) [inline swift/Sources/CoreAILanguageModels/Handlers/InputHandler.swift]: Same fixed-dim issue as register — resolvingDynamicDimensions can't substitute into a fixed static dimension

File: swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAISequentialEngine.swift [hunk 3]
```diff
@@ -42,28 +42,19 @@ public final class CoreAISequentialEngine: InferenceEngine, @unchecked Sendable
     private let functionDescriptor: InferenceFunctionDescriptor
 
     // I/O names from descriptor
-    private let inputIdsName: String
-    private let positionIdsName: String
     private let logitsName: String
 
+    // Input handling — handler owns allocation and fill logic
+    private var inputHandler: TokenInputHandler
```
Comment (carinapeng) [inline swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAISequentialEngine.swift:52]: Another blocker

The sequential engine uses TokenInputHandler, so the new TokenStaticInputHandler which you are adding is unused.

Either wire the sequential path to TokenStaticInputHandler so we get in place fill, or drop it

File: swift/Sources/CoreAILanguageModels/Handlers/InputHandler+Token.swift [hunk 4]
```diff
@@ -97,3 +97,53 @@ public struct CompositeInputHandler<Base: SyncInputHandler>: SyncInputHandler {
...
+
+    public func registerBuffers(into buffers: inout InputBuffers) {
+        buffers.register(name: inputIdsName, descriptor: inputIdsDescriptor)
+        buffers.register(name: positionIdsName, descriptor: positionIdsDescriptor)
+    }
+
+    public func fill(_ context: InputContext, into buffers: inout InputBuffers) throws {
+        let tokens = context.tokens
+        let batchSize = tokens.count
+        precondition(batchSize > 0, "TokenStaticInputHandler: empty token batch")
```
Comment (carinapeng) [inline swift/Sources/CoreAILanguageModels/Handlers/InputHandler+Token.swift]: prefer throwing InferenceRuntimeError over a precondition that crashes on an empty batch

File: swift/Tests/LanguageModelsTests/InputFillerTests.swift
Comment (carinapeng) [inline swift/Tests/LanguageModelsTests/InputFillerTests.swift:1]: The unit tests build InputBuffers from hand-made NDArray(shape:) buffers, so they never exercise register/ensureCapacity with a real static descriptor — which is exactly the path that crashes

A test that registers from an actual descriptor would have caught the crash before device testing, please add one

**Review by carinapeng — CHANGES_REQUESTED** [review]
There are two blockers, suggest testing more and testing end to end before marking as review ready

For perf/CoW, I agree with you that maybe we want to switch to _modify

File: swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAISequentialEngine.swift (hunk 3 above)
Comment (stikves) [reply swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAISequentialEngine.swift:52]: We now moved to `StaticBucketInputFiller`, the other design had one unavoidable copy

File: swift/Tests/LanguageModelsTests/InputFillerTests.swift
Comment (stikves) [reply swift/Tests/LanguageModelsTests/InputFillerTests.swift:1]: Unfortunately `NDArrayDescriptor` initializer is private, and there is no way to properly test this.

Any "hack" we do would be just deceptive as we would only test helpers, not actual descriptors.

File: swift/Sources/CoreAILanguageModels/Handlers/InputHandler.swift (hunk 1 above)
Comment (stikves) [reply swift/Sources/CoreAILanguageModels/Handlers/InputHandler.swift]: Yes, we shouldn't have dynamic shapes in static engine. This got mixed in from the sequential one.

File: swift/Sources/CoreAILanguageModels/Handlers/InputHandler+Token.swift (hunk 4 above)
Comment (stikves) [reply swift/Sources/CoreAILanguageModels/Handlers/InputHandler+Token.swift]: This entire code is now deleted

File: swift/Sources/CoreAILanguageModels/Handlers/InputHandler.swift (hunk 2 above)
Comment (stikves) [reply swift/Sources/CoreAILanguageModels/Handlers/InputHandler.swift]: removed like the other

**Comment by stikves** [comment]
> There are two blockers, suggest testing more and testing end to end before marking as review ready
> 
> For perf/CoW, I agree with you that maybe we want to switch to _modify

They should be handled now, please take a look

File: swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAIStaticShapeEngine.swift [hunk 5]
```diff
@@ -124,6 +128,50 @@ public final class StaticShapeEngine: InferenceEngine, @unchecked Sendable {
...
+            $0.contains("step") && !$0.contains("pos")
+        })
+
+        // Build per-bucket descriptors from ALL extend/prompt functions
+        var bucketDescs: [StaticBucketInputFiller.BucketKey: StaticBucketInputFiller.BucketDescriptors] = [:]
+        for name in extendFunctionNames {
+            let desc = try Self.requireDescriptor(model: model, functionName: name)
+            let ql = Self.queryLength(descriptor: desc, functionName: name)
+            // Parse context length from function name (extend_256_64 → 256)
+            let nameParts = Array(name.split(separator: "_").suffix(2))
```
Comment (carinapeng) [inline swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAIStaticShapeEngine.swift]: This context-length parsing is reimplemented 4 times with diff fallbacks. Let's extract one helper function and use it everywhere this way we can have one helper implemented and tested and not having different divergence at runtime

File: swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAIStaticShapeEngine.swift [hunk 6]
```diff
@@ -124,6 +128,50 @@ public final class StaticShapeEngine: InferenceEngine, @unchecked Sendable {
...
 
+        // Discover input names from the largest-context function
+        let posName = largestExtendDescriptor.inputNames.first(where: { $0.contains("pos") })
+        guard let positionIdsName = posName else {
+            throw InferenceRuntimeError.invalidState("No position_ids input found in model descriptor")
+        }
+        let maskName: String? =
+            largestExtendDescriptor.inputNames.contains("causal_mask")
+            ? "causal_mask" : nil
+        let stepInputName: String? = largestExtendDescriptor.inputNames.first(where: {
```
Comment (carinapeng) [inline swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAIStaticShapeEngine.swift]: This is fragile in how we match input names with `contains()`. Preferred to use exact names like how we do for `causal_mask` or even better a centralized name resolution step

File: swift/Sources/CoreAILanguageModels/Handlers/InputHandler.swift [hunk 7]
```diff
@@ -88,3 +94,107 @@ public enum InputCoverage {
         }
     }
 }
+
+// MARK: - New InputHandler Protocol (inout fill pattern)
+
+/// Pre-allocated input buffer set. Owns NDArrays, reused across steps.
+/// The engine creates this once at init and passes it to `StaticInputHandler.fill()` each step.
+public struct InputBuffers {
```
Comment (carinapeng) [inline swift/Sources/CoreAILanguageModels/Handlers/InputHandler.swift:106]: `InputBuffers` still seems to expose COW cost? A future caller doing a `let a = buffers[name]` would reintroduce the +1 refcount

If they have to stay public, we could add warning of `get` not just on `_modify`

File: swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAIStaticShapeEngine.swift [hunk 8]
```diff
@@ -471,8 +494,21 @@ public final class StaticShapeEngine: InferenceEngine, @unchecked Sendable {
...
+
+        // Fill position_ids, causal_mask, step via the handler
+        let context = InputContext.static(
+            tokens: ArraySlice(batchTokens),
+            alignedStep: alignedStep,
+            batchSize: batchSize,
+            slidingWindow: nil,
+            contextBucket: contextLength)
+        try inputFiller.fill(context, into: &inputBuffers)
+        var inputs = inputBuffers.asDict()
```
Comment (carinapeng) [inline swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAIStaticShapeEngine.swift]: `asDict()` copies the dict every step and the correctness leans on it not being retained?

File: swift/Sources/CoreAILanguageModels/Handlers/InputHandler+StaticBucket.swift [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        // Causal mask: [1, ctx, 1, batch] — lower-triangular
+        if let maskName = causalMaskName, let maskDesc = descs.causalMask {
+            buffers.ensureCapacity(name: maskName, descriptor: maskDesc)
+            try buffers.withMutableBuffer(maskName) { array in
+                array.mutableView(as: Float16.self)
+                    .withUnsafeMutablePointer { ptr, shape, strides in
+                        for ctx in 0..<shape[1] {
+                            for query in 0..<shape[3] {
+                                let offset = ctx &* strides[1] &+ query &* strides[3]
+                                ptr[offset] = Float16(-40000)
```
Comment (carinapeng) [inline swift/Sources/CoreAILanguageModels/Handlers/InputHandler+StaticBucket.swift]: Minor: consider a named constant shared w the dynamic path so we don't drift between engines

File: swift/Sources/CoreAILanguageModels/Handlers/InputHandler+StaticBucket.swift [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+                buffers.preAllocate(name: sName, descriptor: sDesc)
+            }
+        }
+    }
+
+    public func fill(_ context: InputContext, into buffers: inout InputBuffers) throws {
+        let batchSize = context.batchSize
+        let alignedStep = context.alignedStep
+        let tokensInBatch = context.tokens.count
+        let contextLength = context.contextBucket
```
Comment (carinapeng) [inline swift/Sources/CoreAILanguageModels/Handlers/InputHandler+StaticBucket.swift:74]: Nit: we recompute `contextLength = context.contextBucket` and look up the bucket every step, we can make this more efficient by passing a resolved descriptor at graph selection time

File: swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAIStaticShapeEngine.swift (hunk 5 above)
Comment (stikves) [reply swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAIStaticShapeEngine.swift]: Added `parseFunctionDimensions`

File: swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAIStaticShapeEngine.swift (hunk 6 above)
Comment (stikves) [reply swift/Sources/CoreAILanguageModels/InferenceEngines/CoreAIStaticShapeEngine.swift]: `resolveInputName(from:candidates:)` with explicit list

File: swift/Sources/CoreAILanguageModels/Handlers/InputHandler.swift (hunk 7 above)
Comment (stikves) [reply swift/Sources/CoreAILanguageModels/Handlers/InputHandler.swift:106]: Added doc comments, we should not pollute output

**Review by carinapeng — APPROVED** [review]
Let's do more testing before merging :)

**Comment by stikves** [comment]
Tested both correctness and performance

This new handler improves existing generation by 0.9% to 7.5% depending on the model and the chipset (by avoiding unnecessary copies)

Will merge after rebase.
